src/main/java/org/testng/internal/MethodRunner.java

```diff
@@ -0,0 +1,133 @@
+      boolean skipFailedInvocationCounts) {
+    List<ITestResult> result = Lists.newArrayList();
+    int parametersIndex = 0;
+    while (allParameterValues.hasNext()) {
+      Object[] next = allParameterValues.next();
+      if (next == null) {
+        // skipped value
+        parametersIndex++;
+        continue;
+      }
+      Object[] parameterValues =
+          Parameters.injectParameters(
+              next, arguments.getTestMethod().getConstructorOrMethod().getMethod(), context);
+
+      List<ITestResult> tmpResults = Lists.newArrayList();
+      int tmpResultsIndex = -1;
+      TestMethodArguments tmArguments = new Builder()
+          .usingArguments(arguments)
+          .withParameterValues(parameterValues)
+          .withParametersIndex(parametersIndex)
+          .build();
+      try {
+        ITestResult tmpResult = testInvoker
+            .invokeTestMethod(tmArguments, context.getSuite().getXmlSuite(), failure);
+        tmpResults.add(tmpResult);
+        tmpResultsIndex++;
+      } finally {
+        boolean lastSucces = false;
+        if (tmpResultsIndex >= 0) {
+          lastSucces = (tmpResults.get(tmpResultsIndex).getStatus() == ITestResult.SUCCESS);
+        }
+        if (failure.instances.isEmpty() || lastSucces) {
+          result.addAll(tmpResults);
+        } else {
+          List<ITestResult> retryResults = Lists.newArrayList();
+          failure = testInvoker.retryFailed(tmArguments, retryResults, failure.count, context);
+          result.addAll(retryResults);
+        }
+
+        // If we have a failure, skip all the
+        // other invocationCounts
+        if (failure.count > 0
+            && (skipFailedInvocationCounts || tmArguments.getTestMethod()
+            .skipFailedInvocations())) {
+          while (invocationCount.getAndDecrement() > 0) {
+            ITestResult r = testInvoker
+                .registerSkippedTestResult(tmArguments.getTestMethod(), System.currentTimeMillis(),
+                    null);
+            result.add(r);
+            InvokedMethod invokedMethod =
+                new InvokedMethod(r.getInstance(), tmArguments.getTestMethod(),
+                    System.currentTimeMillis(), r);
+            testInvoker.invokeListenersForSkippedTestResult(r, invokedMethod);
+          }
+        }
+      } // end finally
+      parametersIndex++;
+    }
+    return result;
+  }
+
+  @Override
+  public List<ITestResult> runInParallel(TestMethodArguments arguments,
+      ITestInvoker testInvoker,
+      ITestContext context,
+      AtomicInteger invocationCount,
+      FailureContext failure,
+      Iterator<Object[]> allParameterValues,
+      boolean skipFailedInvocationCounts) {
+    XmlSuite suite = context.getSuite().getXmlSuite();
+    List<ITestResult> result = Lists.newArrayList();
+    List<TestMethodWithDataProviderMethodWorker> workers = Lists.newArrayList();
+    int parametersIndex = 0;
+    while (allParameterValues.hasNext()) {
```

Comment:
Use foreach here

Reply:
Cannot use `foreach` here since `allParameterValues` is an iterator.

Reply:
Rigth! Will it be a lot of work to replace the Iterator by an Iterable?

Reply:
Fixed.